### PR TITLE
[compiler/circle-input-names] Introduce a new tool displaying input names

### PR DESCRIPTION
**Pull request by batcheu**
The new tool 'circle_input_names' displays the all input names of each operator in JSON format.
This will be used to generate input metadata in Model Explorer.

---

* draft : #14458
* for : #14427

### CODE REVIEW

File: compiler/circle-input-names/src/circle-input-names.cpp [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+template <> struct is_VariadicArity<CircleIf> : std::true_type {};
+template <> struct is_VariadicArity<CircleWhile> : std::true_type {};
+
+// For Variadic Outputs Nodes which have multiple outputs.
+template <typename T> struct is_VariadicOut : std::false_type {};
+template <> struct is_VariadicOut<CircleCustom> : std::true_type{};
+template <> struct is_VariadicOut<CircleIf> : std::true_type {};
+template <> struct is_VariadicOut<CircleWhile> : std::true_type {};
+// clang-format on
+
```
Comment (batcheu) [inline compiler/circle-input-names/src/circle-input-names.cpp:39]: ### Note
Here, SFINAE is used to create CircleNode depending on default contructor for each.

File: compiler/circle-input-names/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+file(GLOB SOURCES "src/*.cpp")
+
+add_executable(circle_input_names ${SOURCES})
+target_include_directories(circle_input_names PUBLIC include)
```
Comment (seanshpark) [inline compiler/circle-input-names/CMakeLists.txt:4]: I do not see `include` folder (in also draft). can we remove this line?

**Comment by seanshpark** [comment]
I can see `circle-input-names-test` module in draft.
If possible can we add test codes in this module too, if there is no particular reason to split two.

File: compiler/circle-input-names/CMakeLists.txt (hunk 2 above)
Comment (batcheu) [reply compiler/circle-input-names/CMakeLists.txt:4]: Oh, I'll remove it.
Thanks ;)

**Comment by batcheu** [comment]
> I can see `circle-input-names-test` module in draft. If possible can we add test codes in this module too, if there is no particular reason to split two.

Okay, I'll revise it too.

File: compiler/circle-input-names/src/circle-input-names.cpp [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+  }
+  return;
+}
+
+// Mock Symbol Table for CircleNodeSummaryBuilder
+class MockSymbolTable : public locop::SymbolTable
+{
+  std::string lookup(const loco::Node *) const override { return ""; }
+};
+
```
Comment (batcheu) [inline compiler/circle-input-names/src/circle-input-names.cpp:132]: ### Note

In fact, this tool is nothing but prints all input names defined in `CircleNodeSummaryBuilder`.
However, it requires some hacks to instantiate the builder outside of the module.
All above codes are the **Hacks**.

File: compiler/circle-input-names/src/circle-input-names.cpp [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+int main(void)
+{
+  std::stringstream ss;
+  crew::JsonExport json_export(ss);
+  // "{"
+  json_export.open_brace();
+#define CIRCLE_NODE(OP, CIRCLE_OP)              \
+  {                                             \
```
[the remark this answers is not in the document]
Comment (batcheu) [reply compiler/circle-input-names/src/circle-input-names.cpp:163]: Good point, I'll add a wrapper function as you suggested.
Thanks !

File: compiler/circle-input-names/README.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# circle-input-names
+
+`circle-input-names` is a tool to generate input names of the Circle model's operators in JSON format.
```
[the remark this answers is not in the document]
Comment (batcheu) [reply compiler/circle-input-names/README.md:3]: Good suggestion 2.
I'll update it too.

File: compiler/circle-input-names/src/circle-input-names.test.cpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+#include <cstring>
+#include <fstream>
+#include <vector>
+#include <regex>
+
+class circle_input_names_test : public ::testing::Test
+{
+protected:
+  void SetUp(void) override;
+  std::vector<std::string> get_input_names_of(std::string op);
```
Comment (seanshpark) [inline compiler/circle-input-names/src/circle-input-names.test.cpp:29]: two function name has different convention
```suggestion
  void setup(void) override;
  std::vector<std::string> get_input_names_of(std::string op);
```
or
```suggestion
  void SetUp(void) override;
  std::vector<std::string> GetInputNamesOf(std::string op);
```
?
my preference is the first.

Comment (batcheu) [reply compiler/circle-input-names/src/circle-input-names.test.cpp:29]: The first one is virtual function of GTest's fixture class, so it follows _came case_.
However, in ONE, it seems that it follows _snake case_. 
Which one should I follow it? 🤔

File: compiler/circle-input-names/src/circle-input-names.test.cpp [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  FILE *fp = popen(cmd.c_str(), "r");
+  if (!fp)
+  {
+    std::runtime_error("popen() failed");
+    return;
+  }
+
+  char buff[1024];
```
Comment (seanshpark) [inline compiler/circle-input-names/src/circle-input-names.test.cpp:51]: `1024` fixed length may have potential problem when we long file in the future.
maybe use string. maybe use c++ file instead of `FILE` ?

File: compiler/circle-input-names/src/circle-input-names.test.cpp [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+protected:
+  std::string _input_names_dict_str;
+};
+
+void circle_input_names_test::SetUp(void)
+{
+  std::string cmd = std::getenv("CIRCLE_INPUT_NAMES_PATH");
+  if (cmd.empty())
+  {
+    std::runtime_error("CIRCLE_INPUT_NAMES_PATH is not found");
```
Comment (seanshpark) [inline compiler/circle-input-names/src/circle-input-names.test.cpp:40]: maybe
```suggestion
    throw std::runtime_error("CIRCLE_INPUT_NAMES_PATH is not found");
```
?

File: compiler/circle-input-names/src/circle-input-names.test.cpp [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+TEST_F(circle_input_names_test, valid_names_conv2d)
+{
+  // "CONV_2D" should have three inputs: "input", "filter", "bias"
+  auto names = get_input_names_of("CONV_2D");
+  ASSERT_EQ(names.size(), 3);
+  ASSERT_EQ(names[0], "input");
+  ASSERT_EQ(names[1], "filter");
+  ASSERT_EQ(names[2], "bias");
+}
```
Comment (seanshpark) [inline compiler/circle-input-names/src/circle-input-names.test.cpp:124]: for our SE requirements, we need to add num(positive test) <= num(negative test).
please add some negative tests with name having `_NEG` suffix.

Comment (seanshpark) [reply compiler/circle-input-names/src/circle-input-names.test.cpp:124]: strictly speaking, this(circle-input-names) is not included in our release module,
so negative tests are not counted. :/
but for the quality, negative tests are welcome :)

File: compiler/circle-input-names/src/circle-input-names.test.cpp (hunk 6 above)
Comment (seanshpark) [reply compiler/circle-input-names/src/circle-input-names.test.cpp:29]: I would be better to use same convention, for this second one is preferred.

File: compiler/circle-input-names/src/circle-input-names.test.cpp (hunk 7 above)
Comment (batcheu) [reply compiler/circle-input-names/src/circle-input-names.test.cpp:51]: This temporary local buffer is reused in following line if the file size exceeds the buffer size.
Anyway, I'll find a other way instead of using fixed size of buffer.

File: compiler/circle-input-names/src/circle-input-names.test.cpp (hunk 9 above)
Comment (batcheu) [reply compiler/circle-input-names/src/circle-input-names.test.cpp:124]: Okay, I'll update it.

File: compiler/circle-input-names/src/circle-input-names.test.cpp (hunk 7 above)
Comment (seanshpark) [reply compiler/circle-input-names/src/circle-input-names.test.cpp:51]: I think current code is OK. You may revise if you like :)

File: compiler/circle-input-names/src/circle-input-names.cpp [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+// For Variadic Outputs Nodes which have multiple outputs.
+template <typename T> struct is_VariadicOut : std::false_type {};
+template <> struct is_VariadicOut<CircleCustom> : std::true_type{};
+template <> struct is_VariadicOut<CircleIf> : std::true_type {};
+template <> struct is_VariadicOut<CircleWhile> : std::true_type {};
+// clang-format on
+
+// For Circle Nodes which have variadic arity and variadic outputs
+template <typename CircleOp, typename std::enable_if_t<is_VariadicArity<CircleOp>::value &&
+                                                       is_VariadicOut<CircleOp>::value> * = nullptr>
```
Comment (zetwhite) [inline compiler/circle-input-names/src/circle-input-names.cpp:42]: Template study really helps me understand this code 👍 
But still I have one question; Why `* = nullptr` is necessary here? 

```suggestion
template <typename CircleOp, typename std::enable_if_t<is_VariadicArity<CircleOp>::value &&
                                                       is_VariadicOut<CircleOp>::value>>
```

Comment (zetwhite) [reply compiler/circle-input-names/src/circle-input-names.cpp:42]: Ah, I asked ChatGPT and found an answer! 
Because there are several `template<...> auto CircleCreator()` exist, `*=nullptr` looks necessary.

**Review by zetwhite — COMMENTED** [review]
I read the code carefully. Short and powerful code 👍 
I'd like to leave LGTM after all CI pass.

File: compiler/circle-input-names/src/circle-input-names.test.cpp (hunk 7 above)
Comment (batcheu) [reply compiler/circle-input-names/src/circle-input-names.test.cpp:49]: I've tried to find replacible method, but I failed.
It seems that there's no generic way to read file from `FILE *` to streambuf class.
I just left it as it is for now because current code is not hamful and follows generic use pattern. 
I'll revise if I find a better way.

File: compiler/circle-input-names/src/circle-input-names.test.cpp (hunk 6 above)
Comment (batcheu) [reply compiler/circle-input-names/src/circle-input-names.test.cpp]: Instead of forcing the one convention, I added comments of each member function for explanation of different conventions.

File: compiler/circle-input-names/src/circle-input-names.test.cpp (hunk 8 above)
Comment (batcheu) [reply compiler/circle-input-names/src/circle-input-names.test.cpp]: Applied!

File: compiler/circle-input-names/src/circle-input-names.test.cpp (hunk 9 above)
Comment (batcheu) [reply compiler/circle-input-names/src/circle-input-names.test.cpp:127]: Applied!

File: compiler/circle-input-names/src/circle-input-names.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (batcheu) [reply compiler/circle-input-names/src/circle-input-names.cpp]: Applied!

@mhs4670go
PTAL if you have a spare time :)

File: compiler/circle-input-names/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (batcheu) [reply compiler/circle-input-names/README.md:3]: Applied!

@mhs4670go
ditto.

**Review by seanshpark — APPROVED** [review]
LGTM thank you!

File: compiler/circle-input-names/src/circle-input-names.cpp [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+  // "{"
+  json_export.open_brace();
+#define CIRCLE_NODE(OP, CIRCLE_OP)                             \
+  {                                                            \
+    auto node = CircleNodeCreator<CIRCLE_OP>();                \
+    fill_dummies_for_summary_creation(&node);                  \
+    auto summary = create_circle_node_summary(&node);          \
+    get_input_names_from_summary(&node, summary, json_export); \
+  }
+#define CIRCLE_VNODE(_1, _2)
```
Comment (mhs4670go) [inline compiler/circle-input-names/src/circle-input-names.cpp:183]: Just curiosity, what does this code mean? What does it do?

File: compiler/circle-input-names/src/circle-input-names.test.cpp [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+void circle_input_names_test::SetUp(void)
+{
+  std::string cmd = std::getenv("CIRCLE_INPUT_NAMES_PATH");
+  if (cmd.empty())
+  {
+    throw std::runtime_error("CIRCLE_INPUT_NAMES_PATH is not found");
+  }
+
+  FILE *fp = popen(cmd.c_str(), "r");
```
Comment (mhs4670go) [inline compiler/circle-input-names/src/circle-input-names.test.cpp:48]: (Optional) Just as a minor comment, I think we can revisit the program not to use popen function in the test by making a function that returns the stringstream. Which makes a test flow simpler.

```c++
int main()
{
  #  ..
  auto ss_input_names = # ..
  std::cout << ss_input_names.str();

  return 0;
}
```

But, this program seems not to have high maintainability. So, it's not that important.

**Review by mhs4670go — APPROVED** [review]
LGTM

File: compiler/circle-input-names/src/circle-input-names.cpp (hunk 11 above)
Comment (batcheu) [reply compiler/circle-input-names/src/circle-input-names.cpp:183]: Defined to ignore `CIRCLE_VNODE`s in `circle_nodes.lst`.

File: compiler/circle-input-names/src/circle-input-names.test.cpp (hunk 12 above)
Comment (batcheu) [reply compiler/circle-input-names/src/circle-input-names.test.cpp:48]: If you don't mind, could you elaborate more about it?
Is there a way to redirect forked process standard output to stream buffer?

Comment (batcheu) [reply compiler/circle-input-names/src/circle-input-names.test.cpp:48]: @seanshpark explained about it in offline ;)
Now, I understand what you meant, Thanks 👍

**Comment by seanshpark** [comment]
> I'd like to leave LGTM after all CI pass.

I'll land with this message. :)

File: compiler/circle-input-names/src/circle-input-names.cpp (hunk 11 above)
Comment (mhs4670go) [reply compiler/circle-input-names/src/circle-input-names.cpp:183]: Ah, I got it. This would be better to be described as a comment. Maybe in the next PR.
